File: computer3/Fall Term/Solutions/Class11_Nov_20/lvl3_pursuit_of_knowledge.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	static int n, m, t, dis[][]; static List<Integer> adj[];
	public static void main(String[] args) throws IOException{	
		n = readInt(); m = readInt(); t = readInt(); adj = new ArrayList[n+1];
		dis = new int[n+1][n+1];
		for(int i=1; i<=n; i++) {
			adj[i] = new ArrayList<>(); 
			Arrays.fill(dis[i], -1);
		}
		for(int i=1; i<=m; i++) {
			int u = readInt(), v = readInt(); adj[u].add(v);
		}
		for(int i=1; i<=n; i++) bfs(i, dis[i]);
		for(int q=readInt(); q > 0; q--) {
			int u = readInt(), v = readInt();
			if(dis[u][v] != -1) System.out.println(dis[u][v] * t);
			else System.out.println("Not enough hallways!");
		}
	}
	static void bfs(int s, int dis[]) {
		Queue<Integer> q = new LinkedList<>();
		q.add(s); dis[s] = 0;
		while(!q.isEmpty()) {
			int u = q.poll();
			for(int v : adj[u]) {
				if(dis[v]==-1) { q.add(v); dis[v] = dis[u] + 1; }
			}
		}
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: computer3/Fall Term/Solutions/Class11_Nov_20/lvl3_pursuit_of_knowledge.cpp
#include <bits/stdc++.h>
using namespace std;
int N, M, T, Q; vector<int> adj[1001];
int bfs(int st, int ed){
    queue<int> q; bool vis[1001]; int dis[1001];
    memset(vis, 0, sizeof(vis));  // fill(vis, vis+1001, 0);
    q.push(st); vis[st]=1; dis[st]=0;
    while(!q.empty()){
        int cur = q.front(); q.pop();
        for(int nxt: adj[cur]){
            if(!vis[nxt]){
                q.push(nxt); vis[nxt] = 1; dis[nxt] = dis[cur] + 1;
            }
            if(vis[ed]) return dis[ed];
        }
    }
    return -1;
}
int main(){
    scanf("%d %d %d", &N, &M, &T);
    for(int i=1, u, v; i<=M; i++){
        scanf("%d %d", &u, &v);
        adj[u].push_back(v);
    }
    scanf("%d", &Q);
    for(int i=1, a, b; i<=Q; i++){
        scanf("%d %d", &a, &b);
        int ans = bfs(a, b) * T;
        if(ans < 0) printf("Not enough hallways!\n");
        else printf("%d\n", ans);
    }
}